api/topic.go

```diff
@@ -65,6 +62,116 @@ func (api *API) getTopicPrivateHandler(w http.ResponseWriter, req *http.Request)
 	log.Event(ctx, "request successful", log.INFO, logdata) // NOTE: name of function is in logdata
 }
 
+// getSubtopicsPublicHandler is a handler that gets a topic by its id from MongoDB
+func (api *API) getSubtopicsPublicHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	vars := mux.Vars(req)
+	id := vars["id"]
+	logdata := log.Data{
+		"request_id": ctx.Value(dprequest.RequestIdKey),
+		"topic_id":   id,
+		"function":   "getSubtopicsPublicHandler",
+	}
+
+	// get topic from mongoDB by id
+	topic, err := api.dataStore.Backend.GetTopic(id)
+	if err != nil {
+		// no topic found to retrieve the subtopics from
+		handleError(ctx, w, err, logdata)
+		return
+	}
+
+	// User is not authenticated and hence has only access to current sub document(s)
+	var result models.PublicSubtopics
+
+	if topic.Current == nil {
+		handleError(ctx, w, apierrors.ErrInternalServer, logdata)
+		return
+	}
+
+	if len(topic.Current.SubtopicIds) == 0 {
+		// no subtopics exist for the requested ID
+		handleError(ctx, w, apierrors.ErrNotFound, logdata)
+		return
+	}
+
+	for _, subTopicID := range topic.Current.SubtopicIds {
+		// get sub topic from mongoDB by subTopicID
+		topic, err := api.dataStore.Backend.GetTopic(subTopicID)
+		if err != nil {
+			logdata["missing subtopic for id"] = subTopicID
+			log.Event(ctx, err.Error(), log.ERROR, logdata)
+			continue
+		}
+		result.PublicItems = append(result.PublicItems, topic.Current)
+		result.TotalCount++
+	}
+	if result.TotalCount == 0 {
+		handleError(ctx, w, apierrors.ErrInternalServer, logdata)
+		return
+	}
+
+	if err := WriteJSONBody(ctx, result, w, logdata); err != nil {
+		return
+	}
+	log.Event(ctx, "request successful", log.INFO, logdata) // NOTE: name of function is in logdata
+}
+
+// getSubtopicsPrivateHandler is a handler that gets a topic by its id from MongoDB
+func (api *API) getSubtopicsPrivateHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	vars := mux.Vars(req)
+	id := vars["id"]
+	logdata := log.Data{
+		"request_id": ctx.Value(dprequest.RequestIdKey),
+		"topic_id":   id,
+		"function":   "getSubtopicsPrivateHandler",
+	}
+
+	// get topic from mongoDB by id
+	topic, err := api.dataStore.Backend.GetTopic(id)
+	if err != nil {
+		// no topic found to retrieve the subtopics from
+		handleError(ctx, w, err, logdata)
+		return
+	}
+
+	// User has valid authentication to get raw full topic document(s)
+	var result models.PrivateSubtopics
+
+	if topic.Next == nil {
```

Comment:
Could it be valid to have a nil next document? If a topic is published, and no other topic is due to be published, the next doc could be nil?

Reply:
The next doc is always populated, but when its been published should be exactly the same as the current (following the current Dataset API pattern, we should be copying 'next' over 'current' when the state is set to 'published'.

Because of this, it's valid for current to be nil (its never been published) but never next.